File: 백준/Platinum/3015. 오아시스 재결합/오아시스 재결합.py
import math
import sys
input = sys.stdin.readline
n = int(input())
stack = []
result = 0
for i in range(n):
    a = int(input())
    while True:
        if not stack or stack[-1][0]>a:
            stack.append([a,1])
            break
        elif stack[-1][0]==a:
            stack[-1][1]+=1
            break
        else:
            _,c=stack.pop()
            result+=(math.comb(c,2)+c)
            if stack:
                result+=c
    # print(stack,result)

while stack:
    _,c = stack.pop()
    result+=math.comb(c,2)
    if stack:
        result+=c
print(result)

# # 2    4    1    2    2    5    1
# # 제일 작은 수 +1
# # 제일 큰 수 +len
# # 중간 수 제거+1

# # 2    4    [4]
# # 4    1    [4,1]
# # 4    2    [4,2]
# # 1    2
# # 4    2    [4,2,2]
# # 2    2
# # 4    5    [5]
# # 2    5
# # 2    5
# # 5    1    [5,1]

File: 백준/Platinum/3015. 오아시스 재결합/오아시스 재결합.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        Stack<Person> stack = new Stack<>();
        long result = 0;

        for (int i = 0; i < n; i++) {
            int height = Integer.parseInt(br.readLine());
            while (true) {
                if (stack.isEmpty() || stack.peek().height > height) {
                    stack.push(new Person(height, 1));
                    break;
                } else if (stack.peek().height == height) {
                    stack.peek().count++;
                    break;
                } else {
                    Person popped = stack.pop();
                    result += (long)popped.count * (popped.count - 1) / 2 + popped.count;
                    if (!stack.isEmpty()) result += popped.count;
                }
            }
        }

        while (!stack.isEmpty()) {
            Person p = stack.pop();
            result += (long)p.count * (p.count - 1) / 2;
            if (!stack.isEmpty()) result += p.count;
        }

        System.out.println(result);
    }

    static class Person {
        int height;
        int count;

        Person(int height, int count) {
            this.height = height;
            this.count = count;
        }
    }
}
